### ISSUE TITLE
Local Orders visible to Non-members in Timesheet view

### ISSUE THREAD

**Issue by greyman888** [maxrossello/redmine_app_timesheets#70]
Hi Max,

Build: timesheets 1.5.5, Redmine 3.1, PostgreSQL & MySQL
Config: "Shared versions visible to non members" set to nil

**Example Setup**
In a project called Project1 with the settings:
Public: true
Members: user1
Versions: version1 with "Is order" set to true and "Sharing" set to "Not shared"

**Action**
When a second user (user2) who is not a member of Project1 accesses their timesheet, version1 is available in the Order selector. The user should not be able to put time against this order (version1) because they are not a member of the project.

This bug has always been there but in the past I have made sure that all projects that stored orders had the setting "Public" set to false.

With the changes to timesheet >1.4 any user can create an order in any project and so I've lost the control I previously had to limit access to orders (previously admins had to share versions with all projects to enable them as orders).

As we have lots of orders, lots of users and lots of projects, it will quickly get messy and confusing if users start seeing orders that they're not meant to use.

I have tested a change to timesheets_controller.rb that affects the lines below in the get_user method. The existing lines are:

    WorkOrder.not_native.all.each do |order|
       @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
       @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
       @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
       @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
       @visibility[order.id] ||= TsPermission::FORBIDDEN
    end

The change is:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      if Setting.plugin_redmine_app_timesheets["public_versions"].nil?
        @visibility[order.id] = TsPermission::FORBIDDEN if order.project.members.find_by_user_id(User.current).nil?
      end
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
    end

All it's doing is setting the permission to FORBIDDEN if the user is not a member of the project. On the next line, the permission is set to ADMIN if the user is an admin (even if they're not a member). I have kept this in place as admins see everything everywhere else.

I would have raised this as a pull request but since I already have one open, I couldn't work out how to make a new one for this change only.

Also, while I was doing some testing I found these lines in the get_timelogs method:

    # in add row, add orders where I am admin
    @manageable_orders << WorkOrder.enabled.where("id IN (?)", @visibility.select{|k,v| v == TsPermission::ADMIN}.keys )

With PostgreSQL at least, it seems to be taking the @manageable_orders array and then appending an active record relationship which means the array now contains mixed types.

I changed the line to:

    if orders_where_user_is_admin = WorkOrder.enabled.find_by_id(@visibility.select{|k,v| v == TsPermission::ADMIN}.keys)
       @manageable_orders << orders_where_user_is_admin
    end

So that the now the array @manageable_orders contains elements of all the same type and to stop a nil being appended. 

Unfortunately I couldn't find any use cases where this line was appending any orders that weren't there already. So really my question is, does this line need to be there at all?

Adam

**Comment by maxrossello** [comment]
Hi Adam,
for the first part: I think that a non member would fall back to FORBIDDEN unless he has permissions in the project or is an admin. Can you report a use case where this is not true? Maybe the @visibility needs a fallback pre-setting to FORBIDDEN _before_ the checks (maybe it's not nil at the beginning)?
I also think that making it dependent from "public_versions" is pointless.

The second: ok

Sorry for being less responsive in importing changes, these are very hectic times.

Thanks a lot

**Comment by greyman888** [comment]
Hi Max, thanks for the reply. I appreciate how busy you are.

This is a complicated problem so please forgive me if I have misunderstood something somewhere. The use case I'm working on is the same as I described in the issue. Any non-admin user who not a member of the project that contains an order. I do not want them to see the order in the "Add row" selector because I am using project membership to control access to orders.

* In Redmine Administration - Roles and permissions, you can set the permissions based on a role.
* The role I am concerned with is *Non member*
* In a new install of Redmine the default permission is that non-members can "View spent time"
* This permission is set either for all projects or for none, I am not aware of per project permission adjustment

If I deselect "View spent time", then back in the timesheet the non-member does not see the order in the "Add row" selector. This is good but still indicates that there is a bug because being able to view a time log should not interfere with being able to add a new one. 

>What makes it even more complex is that I have inherited a production configuration where the *Non member* permissions are set to Log spent time, View spent time & Edit own time logs. This isn't great and undermines the point of permissions but it reflects the fact that within the organisation people from different teams sometimes collaborate. Now that I better understand how the timesheet works, I will remove these permissions from the *Non-member* role and ensure that users must be members to log time to a project. I will include some of this in the wiki.

So back to this issue, in the get_user method, the lines below are applying the correct TsPermissions and no change is required.

      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN

It is in the get_timelogs method where these lines determine what orders the user can select:

      if order.in_timesheet
        if  ((User.current == @user and @visibility[order.id] != TsPermission::FORBIDDEN) or
            @visibility[order.id] >= TsPermission::EDIT)
          @manageable_orders << order
        end
      end

It seems that if TsPermission is VIEW (1) then the first condition on the if statement is true and so the order is included for selection.

I know it's not pretty but a change to this worked:

      if order.in_timesheet
        if  User.current == @user and 
          @visibility[order.id] != TsPermission::FORBIDDEN and
          @visibility[order.id] != TsPermission::VIEW
          @manageable_orders << order
        end
      end

What do you think?

**Comment by maxrossello** [comment]
First part: I see, I usually set no privileges to non-member because I find it too unsafe and pointless to have a single setting for all projects. I think there is some plugin out there to fix this weak point.
However, the timesheet acts accordingly to redmine settings so yes, I find no change is necessary.

 > This is good but still indicates that there is a bug because being able to view a time log should not interfere with being able to add a new one. 

I think you are right. Former time entries should be displayed read-only and there is no use to have the order displayed in the drop down. This should be true for every member type.

The problem with your last proposal is that a manager wouldn't be able to edit a user timesheet due to "User.current == @user *and* ".
I need to double check if @manageable_orders is not used elsewhere (for example to select the displayed rows, which should be seen read-only in VIEW mode), but in that case then it would be sufficient to keep the order open to visibility >= TsPermission::EDIT for any user, since we are saying there is no point to keep it listed in VIEW mode for the current user looking at his own timesheet.

**Comment by greyman888** [comment]
Sorry I missed the order manager problem. I can not see any other uses of `@manageable_orders` so I just tested the simplification you suggested:
    
    if order.in_timesheet
        @manageable_orders << order if  @visibility[order.id] >= TsPermission::EDIT
    end

I think this solution will work and it keeps separate the control of what an order manager can view from what an order manager can edit in another users timesheet. What a manager can view in the timesheet depends on the Redmine permissions for a *non-member*. What a manager can edit depends on their membership of the applicable projects.

If *non-member* is permitted to "View spent time" (default Redmine) then: 

* An order manager can view the timesheet of any user and see all the timelog entries, but they will only be able to add & edit orders where they are also a member of the project.

If *non-member* is **not** permitted to "View spent time" then:

* An order manager can view the timesheet of any user but will not see the timelog entries for projects that they are not also a member of.

The other option is to try and override the Redmine permissions behaviour to always allow an order manager to see and edit another users timesheet whether they have membership of the project or not. I'm less comfortable with this (from a business perspective) and it will require changes to the initial code we discussed in the get_user method. This level of control is retained by Redmine administrators.

**Comment by maxrossello** [comment]
Ok, I like this enhancement.
Agreed, I would not override Redmine's behavior as far as possible, and also I would rather unclutter the managers view, I followed this policy also for native orders.
I hope I can run my tests and import the changes within the weekend. Sorry for the delay

**Comment by greyman888** [comment]
Hi Max,

While writing the guide to put on the wiki I found an unwanted side effect of this change.

The change means that native orders with `TsPermission::NONE` are now no longer available for selection in "Add row". This means that Global Orders do not work if you are not an order admin.

This addition works, but I'm not sure if there is a better fix.

    if @visibility[order.id] >= TsPermission::EDIT or @visibility[order.id] == TsPermission::NONE  
        @manageable_orders << order
    end

Please let me know if I can do anything to help,

**Comment by maxrossello** [comment]
I think we did a mess! :-)
Actually, permission VIEW means that an user is able to EDIT his own entries (so entry in add view shall be visible), and VIEW the other's. This is why the current user had the possibility to view the entry even with NONE (which means can edit only his own).

I'll therefore revert the modifications.

**Comment by maxrossello** [comment]
Reverted

**Comment by greyman888** [comment]
U

**Comment by greyman888** [comment]
Unfortunately that still leaves the same problem that caused me to open the issue in the first place.

**Comment by greyman888** [comment]
If the permission is VIEW then the order shouldn't be available in the 'Add rows' selector otherwise non-members of a project who have the permission to view time entries (default Redmine) will be able to add codes to their timesheet that they shouldn't have permission for.

**Comment by maxrossello** [comment]
Can you please reword it? Our previous exchanges are not pretty clear now, as we went through some understanding.
I think these settings should work also with non-members provided that related permissions are set properly.
If the issue is due to public projects, then I guess that Redmine access should not be touched. All the visibility depends on User.current.allowed_to?

**Comment by maxrossello** [comment]
We are writing in parallel... I see

**Comment by maxrossello** [comment]
Ok, I guess we should use @visibility only for access of foreign timesheets then. I completely missed the non-member part.

I would reframe as such:

<pre>
if  ((User.current == @user and User.current.allowed_to? :edit_own_time_entries, order.project) or
        @visibility[order.id] >= TsPermission::EDIT)
</pre>

can you try it?

**Comment by greyman888** [comment]
I will try it now.

**Comment by greyman888** [comment]
OK. This is really complex so please be patient with me. :-)

The line below works fine for Local Orders (not_native):

    @manageable_orders << order if  @visibility[order.id] >= TsPermission::EDIT

The problem is when Global Orders (native) have the `TsPermission::NONE` (0) so they are skipped over. This is the default when I add a group or individual user to a Global Order who are not admins or order managers.

I tried the line below that you asked me to test and it did not work.

    if  ((User.current == @user and User.current.allowed_to? :edit_own_time_entries, order.project) or
       @visibility[order.id] >= TsPermission::EDIT)

This line is trying to test if the current user has the permission `edit_own_time_entries` of the backing workspace project. When I configure the timesheet, I do not add all users as members to the backing workspace so this condition will be false.

Although I do not completely understand the lines below from the get_user method, it consistently selects the Global Orders that the user has permission to and sets to the permission to `NONE`.

    @visibility = TsPermission.for_user.inject({}) do |h,v|
      h[v[:order_id]] = v[:access] if h[v[:order_id]].nil? or h[v[:order_id]] < v[:access]
      h
    end

So because the lines above are selecting the correct Global Orders and their permission is 'NONE' then that explains why the lines below are working:

    if @visibility[order.id] >= TsPermission::EDIT or @visibility[order.id] == TsPermission::NONE
       @manageable_orders << order
    end

The first part of the if statement picks up Local and Global orders that the user has permission to edit in anyone's timesheet and the second part of the if statement picks up Global Orders that the user has permission to edit only in their own timesheet.

I have tried the following specific tests:

1. Create a Global Order and give permission to User1 but not User2
Result: User1 can select the Global Order but User2 can not

2. Create a Global Order and give permission to User1 and view permission to User2
Result: User2 can only view the User1 time entries for that order and can not select the order in their own timesheet.

3. Create a Local Order and make User1 a member of the project but not User2
Result: User1 can select the Local Order but User2 can not

4. Create a Local Order and make User1 a member of the project and 
make User2 an Order Manager but not a member of the project and 
set Role Permission for Non member 'View spent time' to `true`.
Result: User2 can see the time entries of User1 but can not edit them or select the Local Order

5. Same as above but set Role Permission for Non member 'View spent time' to `false`.
Result: User2 cannot see the Local Order time entries of User1.

I have also tried several other combinations and all seems to be working fine.

**Comment by maxrossello** [comment]
Ok, true. The proposed patch worked only with local orders.

I think you proposal would fail if non-member would be allowed to edit its own time entries in the project (for example if a previous member of a group has been expunged, yet has previous timelogs that he may edit, and also he is granted to view at time entries so @visibility is VIEW).

The fact is that how @visibility is build for local orders comes from the assumption that VIEW is a superset of NONE, while it may not be true.

A complete solution would be to treat permissions as independent ones with bit values (e.g. NONE=1, VIEW=2, EDIT=4, ADMIN=8) and then test the bits. It may be overkill and need a deeper review.

A workaround would be to test native and local orders differently, so 

    (User.current == @user and User.current.allowed_to? :edit_own_time_entries, order.project)

is used for local orders and former one

    (User.current == @user and @visibility[order.id] != TsPermission::FORBIDDEN)

for global ones, while keeping it open on >= EDIT for both

**Comment by greyman888** [comment]
It's pretty late here in Melbourne so I will have to sleep on it.

Tomorrow I will try the scenario you described.

I will add a user as a member to a project, enter some time and then remove
the member from the project. I will set the non-member permission to
edit_own_time_logs.

In that case I think the correct outcome should be that the member can view
and edit their past time logs but should not be able to select the order to
add it as a new row in the timesheet.

I'm having trouble seeing why my proposed change won't work because it only
affects what appears in the 'Add rows' selector.

If you could post a few more scenarios that you would like me to test I
would be happy to. Right now I've tested everything that we use on a daily
basis and it all passes.

**Comment by maxrossello** [comment]
This is just a corner case for overall consistency. 
I think that for displaying the add row entry the 'Log spent time' permission would be actually more suitable to be used than 'Edit own time logs', in order to better cope with Redmine permissions. So for local orders the line would instead better be:

    (User.current == @user and User.current.allowed_to? :log_time, order.project)

Global orders' permissions are instead stored in TsPermission table and we can manage them as we want. Such a discrepancy makes them needing to be treated differently for add_row, because with global orders there is no difference among the 'create' and 'edit' of own entries. For global orders, NONE and, more so, VIEW should be allowed to add entries so the != FORBIDDEN statement is fine.

A couple of scenarios:

6. Create a Local Order and make User1 a member of the project and make User2 not a member of the project and set Role Permission for Non member 'Log Spent Time' to true.

Expected Result: User2 cannot see the time entries of User1 but **should be able** to select the Local Order and add time to it, as he can add time to the Redmine time tracking. Not clear if he can edit existing time.

7. Like the above, but 'Edit Own time entries' is used instead.
I expect User2 can only edit existing time as you said

8. Create a Local Order and make User1 a member of the project and make User2 not a member of the project and set both Role Permissions for Non member 'Log Spent Time' and 'View spent time' to true.

Expected Result: User2 can see the time entries of User1 and **should be able** to select the Local Order and add time to it, as he can add time to the Redmine time tracking.
I expect this scenario to fail without my proposed patch, because User2 will be assigned VIEW, and thus excluded from add_row despite of 'Log Spent Time'

**Comment by greyman888** [comment]
Ok Max. Go and grab a strong coffee and be prepared for a sad story about how I wasted a day. If you don’t have the time, feel free to just read the start up until the heading ‘Sad Story’.

The short version is that I got off track with trying to align Timesheet Permissions with Redmine Permissions for non-members. I first opened this issue because I do NOT want Users to be able to access order codes in the Timesheet if they are not a member of the project (no matter what the Redmine non member permissions are). It is OK for non members to view the timesheet entries of others if they are an Order Manager but they shouldn’t be able to add those orders to any timesheets.

I believe that this was the original intention of your code in the first place. So before I go too much further please let me know if you agree with this principle.

I used my proposed code for all of the tests:

        if @visibility[order.id] >= TsPermission::EDIT or @visibility[order.id] == TsPermission::NONE
          @manageable_orders << order
        end

The additional changes I would like to suggest and then test further are in the `get_user` method:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::NONE if order.project.members.find_by_user_id(User.current)
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
    end

If you want to read my sad story of how I got to this point, it starts here:

## Sad Story
I started testing with my version of Redmine running PostgreSQL.

## The tests in your last post
### First Test
Your test description:
> 6 - Create a Local Order and make User1 a member of the project and make User2 not a member of the project and set Role Permission for Non member 'Log Spent Time' to true.

**Full settings**
Timesheet setting ‘Shared versions visible to non members’: `true`
Non member permission ‘Log spent time: **`true`**
Non member permission ‘Edit own time logs’: `false`
Non member permission ‘View spent time’: `false`
Project setting ‘Public’: `true`
Version setting ‘Sharing’: `Not shared`
Version setting ‘Is order’: `true`
User1: Member - as Manager (with full permissions – Redmine Default)
User2: Non-member

**Your Expectation**
> User2 cannot see the time entries of User1 but **should be able** to select the Local Order and add time to it, as he can add time to the Redmine time tracking. Not clear if he can edit existing time.

**Results**
* In an Issue, User1 can select Target Version and add time - `pass`
* In the Timesheet, User1 can select the Order and add time - `pass`
* In an Issue, User2 can select Target Version and add time - `pass`
* In the Timesheet, User2 can select the Order and add time - **`fail`** (see discussion below)
* In the Timesheet, User2 can NOT see the time entries of User 1 - `pass`
* In the Timesheet, User2 can edit existing time – no result

This test failed because of the lines in the `get_user` method:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
      puts "****Visibility of " + order.name.to_s + " " + order.id.to_s + " is " + @visibility[order.id].to_s
    end

Because User2 had the permission ‘Log_spent_time’ which does not match any of the conditions, it is given the permission of `FORBIDDEN`. So it is excluded from the @manageable_orders.
So I added a new condition:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :log_time, order.project
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
      puts "****Visibility of " + order.name.to_s + " " + order.id.to_s + " is " + @visibility[order.id].to_s
    end

The results with this change were:

* In an Issue, User2 can select Target Version and add time - `pass`
* In the Timesheet, User2 can select the Order and add time - **`pass`**
* In the Timesheet, User2 can NOT see the time entries of User 1 - `pass`
* In the Timesheet, User2 can edit existing time - **`pass`** (They can edit in the timesheet but they can’t view the time in the Issue)

### Second Test
Your test description:
> 7 - Like the above, but 'Edit Own time entries' is used instead.

**Full settings**
Timesheet setting ‘Shared versions visible to non members’: `true`
Non member permission ‘Log spent time: `false`
Non member permission ‘Edit own time logs’: **`true`**
Non member permission ‘View spent time’: `false`
Project setting ‘Public’: `true`
Version setting ‘Sharing’: `Not shared`
Version setting ‘Is order’: `true`
User1: Member - as Manager (with full permissions – Redmine Default)
User2: Non-member

**Your Expectation**
> I expect User2 can only edit existing time as you said

**Results**
* In an Issue, User2 can NOT select Target Version and add time - `pass`
* In the Timesheet, User2 can NOT select the Order and add time - **`fail`** (see discussion below)
* In the Timesheet, User2 can NOT see the time entries of User 1 - `pass`
* In the Timesheet, User2 can edit existing time – **`pass`**

Because User2 had the permission ‘edit _own_time’ which match one of the conditions, it is given the permission of `NONE`. So it is included in @manageable_orders.

So I removed one of the conditions:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :log_time, order.project
      # @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
      puts "****Visibility of " + order.name.to_s + " " + order.id.to_s + " is " + @visibility[order.id].to_s
    end

By removing this condition, the permission for this order in @visibility was set to `FORBIDDEN`

**Results**
* In an Issue, User2 can NOT select Target Version and add time - `pass`
* In the Timesheet, User2 can NOT select the Order and add time - **`pass`**
* In the Timesheet, User2 can NOT see the time entries of User 1 - `pass`
* In the Timesheet, User2 can edit existing time – **`pass`**

The results of this test confirmed for me that the permissions set in @visibility do not impact what can be see and edited in the existing rows of the timesheet. It seems that you have separated these responsibilities well.

### Third Test
Your test description:
> 8 - Create a Local Order and make User1 a member of the project and make User2 not a member of the project and set both Role Permissions for Non member 'Log Spent Time' and 'View spent time' to true.

**Full settings**
Timesheet setting ‘Shared versions visible to non members’: `true`
Non member permission ‘Log spent time: **`true`**
Non member permission ‘Edit own time logs’: `false`
Non member permission ‘View spent time’: **`true`**
Project setting ‘Public’: `true`
Version setting ‘Sharing’: `Not shared`
Version setting ‘Is order’: `true`
User1: Member - as Manager (with full permissions – Redmine Default)
User2: Non-member

**Your Expectation**
> User2 can see the time entries of User1 and should be able to select the Local Order and add time to it, as he can add time to the Redmine time tracking.
I expect this scenario to fail without my proposed patch, because User2 will be assigned VIEW, and thus excluded from add_row despite of 'Log Spent Time'

**Results**
* In an Issue, User2 can select Target Version and add time - `pass`
* In the Timesheet, User2 can select the Order and add time - **`fail`** (see discussion below)
* In the Timesheet, User2 can NOT see the time entries of User 1 - `pass`
* In the Timesheet, User2 can edit existing time – no result

As you correctly expected the test failed because the permission for the order was set to `VIEW’.

So given that I am starting to believe that the responsibilities are well separated in the code, I tried something very different and removed another condition:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :log_time, order.project
      # @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      # @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
      puts "****Visibility of " + order.name.to_s + " " + order.id.to_s + " is " + @visibility[order.id].to_s
    end

* In an Issue, User2 can select Target Version and add time - `pass`
* In the Timesheet, User2 can select the Order and add time - **`pass`**
* In the Timesheet, User2 can NOT see the time entries of User 1 - `pass`
* In the Timesheet, User2 can edit existing time - **`pass`** (They can edit in the timesheet but they can’t edit the time in the Issue)

## The other tests in my last post
As I have made some big changes, I thought I would try some of the previous tests I have performed.
### Forth Test
Test description:
> Create a Local Order and make User1 a member of the project and 
> make User2 an Order Manager but not a member of the project and 
> set Role Permission for Non member 'View spent time' to true.

**Full settings**
Timesheet setting ‘Shared versions visible to non members’: `true`
Non member permission ‘Log spent time: `false`
Non member permission ‘Edit own time logs’: `false`
Non member permission ‘View spent time’: **`true`**
Project setting ‘Public’: `true`
Version setting ‘Sharing’: `Not shared`
Version setting ‘Is order’: `true`
User1: Member - as Manager (with full permissions – Redmine Default)
User2: Non-member and Order Manager

**Expected Result**
> User2 can see the time entries of User1 but can not edit them or select the Local Order

**Results**
* In an Issue, User2 can NOT select Target Version and add time - `pass`
* In the Timesheet, User2 can NOT select the Order and add time - **`pass`**
* In the Timesheet, User2 can see the time entries of User 1 - **`fail`**
* In the Timesheet, User2 can NOT edit the time entries of User 1 – no result

So this changed my perspective about how the code worked and I tried this instead:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :log_time, order.project
      # @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
      puts "****Visibility of " + order.name.to_s + " " + order.id.to_s + " is " + @visibility[order.id].to_s
    end

**Results**
* In an Issue, User2 can NOT select Target Version and add time - `pass`
* In the Timesheet, User2 can NOT select the Order and add time - **`pass`**
* In the Timesheet, User2 can see the time entries of User 1 - **`pass`**
* In the Timesheet, User2 can NOT edit the time entries of User 1 – **`pass`**

# AND THEN I REMEMBERED TWO THINGS
## I RAISED AN ISSUE #65 FOR POSTGRESQL AND

## I DIDN’T WANT THIS BEHAVIOUR IN THE FIRST PLACE!

**I first opened this issue because I didn’t want non-members of a project to have access to an Order in their Timesheet no matter what the Redmine settings are!**

So after I cried for a bit (joking), I went back to first principles.

1.	I only want members of a project to be able to use the orders in that project
2.	I want to allow Order Managers who are Non members to **view** other time entries
3.	I want to allow Order Managers who are also Project Managers with the permission edit_time_logs to be able to edit other user timesheets.

So then I wrote this:

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::NONE if order.project.members.find_by_user_id(User.current)
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
      puts "****Visibility of " + order.name.to_s + " " + order.id.to_s + " is " + @visibility[order.id].to_s
    end

## Final Tests
### Fifth Test
Test description:
This is a repeat of the first test, but I want a different outcome.
Create a Local Order and make User1 a member of the project and make User2 not a member of the project and set Role Permission all timelog permissions to be true.

**Full settings**
Timesheet setting ‘Shared versions visible to non members’: `true`
Non member permission ‘Log spent time: **`true`**
Non member permission ‘Edit own time logs’: **`true`**
Non member permission ‘View spent time’: **`true`**
Project setting ‘Public’: `true`
Version setting ‘Sharing’: `Not shared`
Version setting ‘Is order’: `true`
User1: Member - as Manager (with full permissions – Redmine Default)
User2: Non-member

**Expectation**
> User2 cannot see the time entries of User1 and **should NOT be able** to select the Local Order and add time to it, this is overriding the Redmine time tracking options. This will allow my users to allocate time to Issues on projects when they are non-members and then later select the correct Order code when they add the time to their timesheet.

**Results**
* In an Issue, User2 can select Target Version and add time - `pass`
* In the Timesheet, User2 can NOT select the Order and add time - **`pass`** 
* In the Timesheet, User2 can NOT see the time entries of User 1 - `pass`
* In the Timesheet, User2 can edit existing time – not tested due to bug #65

### Sixth Test
Test description:
> Create a Local Order and make User1 a member of the project and 
> make User2 an Order Manager but not a member of the project and 
> set Role Permission for Non member 'View spent time' to true.

**Full settings**
Timesheet setting ‘Shared versions visible to non members’: `true`
Non member permission ‘Log spent time: **`true`**
Non member permission ‘Edit own time logs’: **`true`**
Non member permission ‘View spent time’: **`true`**
Project setting ‘Public’: `true`
Version setting ‘Sharing’: `Not shared`
Version setting ‘Is order’: `true`
User1: Member - as Manager (with full permissions – Redmine Default)
User2: Non-member and Order Manager

**Expected Result**
> User2 can see the time entries of User1 but can not edit them or select the Local Order

**Results**
* In an Issue, User2 can NOT select Target Version and add time - `pass`
* In the Timesheet, User2 can NOT select the Order and add time - **`pass`**
* In the Timesheet, User2 can see the time entries of User 1 - **`pass`**
* In the Timesheet, User2 can NOT edit the time entries of User 1 – **`pass`**

**Comment by maxrossello** [comment]
Wow Adam, you did a great amount of work here!
I think we should clear up things a little bit more.

### expected behaviour

> I do NOT want Users to be able to access order codes in the Timesheet if they are not a member of the project (no matter what the Redmine non member permissions are). It is OK for non members to view the timesheet entries of others if they are an Order Manager but they shouldn’t be able to add those orders to any timesheets.

> I believe that this was the original intention of your code in the first place. So before I go too much further please let me know if you agree with this principle.

Well, not actually. My intent is to give project managers control of local orders with the (should be intuitive) enforcement of Redmine permissions. I wrote in the Readme:

> For local Orders, visibility settings are those defined in the sourcing project, while for global Orders they are defined into the Order configuration interface.

I see the use case with modifying the standard Redmine permissions for non-members, but I think that it should at least be enforced by an explicit flag in the plugin settings or even at project level.

### about Order Managers

Order Managers are those who can access the order management view, and therefore can activate/deactivate a version as an order. Other than that, being an Order Manager does not bring more permissions over a Local Order. The Project Manager shall be in full control of the Order and cannot know who is elected an Order Manager.
Agreed, policy used with non-members will bring more unexpected behaviour to the Project Manager, but this is a problem that's common to all Redmine aspects, it is a very weak point but at least it should be well known.

So, if a non-member can view timelogs it is because he has permission to view timelogs, not because he is an Order Manager. Again, a change in this behaviour should be explicitly wanted by setting some flag.

### implementation

I suspect that you didn't fully acknowledge what I said yesterday about creation and editing permissions. I'll try to reword.

@visibility is, in first place, to be used in the timesheet (*edit* case). For Global Orders, it **can** be used for add_row as well because I didn't define any difference among create and edit there.
For Local Orders, there is a difference in Redmine permissions, so basically instead of trying to adapt @visibility to take into account both cases yet with an enumeration set that is insufficient, we should rather use @visibility only for the timesheet, and rely on Redmine permissions check for 'create' (add_row).

### proposal

Stated the above, I'll try to reframe my proposal. Please bear with errors as I'm writing this code as example and not testing it yet.

@visibility should have no change, except for possible customized behaviour for non-members. The following correctly defines the permissions for editing own timelogs, view other's and edit other's in the timesheet.

    WorkOrder.not_native.all.each do |order|
      @visibility[order.id] = TsPermission::NONE if User.current.allowed_to? :edit_own_time_entries, order.project
      @visibility[order.id] = TsPermission::VIEW if User.current.allowed_to? :view_time_entries, order.project
      @visibility[order.id] = TsPermission::EDIT if User.current.allowed_to? :edit_time_entries, order.project
      @visibility[order.id] = TsPermission::ADMIN if User.current.admin?
      @visibility[order.id] ||= TsPermission::FORBIDDEN
    end

For what concerns the add_row context, :edit_own_time_entries is not relevant: :log_time is.
So, if no customized behaviour for non-members is enforced, the code to define @manageable_orders should be like:

    if  (
        # current user enabled to add row for local orders if he can log new timelogs
        (!order.is_native? and User.current == @user and User.current.allowed_to? :log_time, order.project) or
        # current user enabled to add row for global orders in his own timesheet
        (order.is_native? and User.current == @user and @visibility[order.id] != TsPermission::FORBIDDEN) or
         # everybody can edit and add rows in any timesheet if having EDIT permission
        @visibility[order.id] >= TsPermission::EDIT
    )
        @manageable_orders << order
    end

### consideration about customized behaviour

If you wish to modify the standard Redmine behaviour for non-members for what concerns the plugin functionality, please bear in mind that there could be side effects.
Even if the add_row does not show an order, there are other ways to workaround this limitation, which should be considered. For example, a non-member can anyway add time to the version as such, as permitted by Redmine, and then add the entry to the timesheet. Therefore, forbidding a non-member to do actions allowed by Redmine would not be so simple.

**Comment by greyman888** [comment]
Thank you for taking the time to give me such a detailed response.

Sorry I didn't pay enough attention to your earlier comments.

I will read through everything you've posted and experiment with the code
you've provided so I better understand.

What you've proposed looks good.

Thanks again.

**Comment by greyman888** [comment]
Hi Max,

I experimented with the code you have proposed and I agree that it is a better alignment to the Redmine permissions than the current code in version 1.5.9 I think that when you have the time you should incorporate this change and close this issue.

If I still feel later that I would like to modify the standard Redmine permissions for non-members, I will open a new issue as an enhancement and explore the idea of enforcing it by an explicit flag in the plugin settings or even at project level as you suggested. For my specific situation, I think that I will be able to work around it for now.

Thanks.

**Comment by maxrossello** [comment]
Do you have any branch where I can merge software that you verified from?

**Comment by greyman888** [comment]
I have submitted a pull request.

Thanks again.

**Comment by maxrossello** [comment]
Thank you so much!
